### DISCUSSION TITLE
[FEATURE REQUEST] Add the ability to ignore certain top-level objects in YAML files.

### DISCUSSION THREAD

**Discussion by dkacar-oradian** [saltstack/salt#67679] [Feature Requests]
**Is your feature request related to a problem? Please describe.**
This is mostly useful for configuration files, but might have a use elsewhere, as well.

I'd like to be able to use YAML's node anchors, but for that I usually need a placeholder which should be ignored by the code which reads that YAML file.

**Describe the solution you'd like**
For example, I'd like to be able to write `salt-cloud`'s map file which looks like this:
```
_ignore: &dry
  foo: bar

profile:
  - host:
      key: val
      <<: *dry
      another: value
```
After loading that file `_ignore` would be ignored instead of causing an error (which it probably does now).

**Describe alternatives you've considered**
It's usually (but not always) possible to switch from YAML document to Python and then place YAML in a string, load it from the Python code, remove the placeholders and serialize back to YAML (or maybe return Python object).

But it's tedious, breaks integration with anything that expects Jinja/YAML (eg. you don't get syntax highliting or auto-identation in your editor because it's a Python string and not YAML any more) and there should be a better way.

**Additional context**

**Please Note**
If this feature request would be considered a substantial change or addition, this should go through a SEP process here https://github.com/saltstack/salt-enhancement-proposals, instead of a feature request.

**Comment by OrangeDog** [comment]
You can achieve the same effect with Jinja wherever it is available.

If you want to use node anchors in raw YAML then you have to use the first occurrence as the reference, not add extra junk at the start.

**Comment by dkacar-oradian** [comment]
To begin with I have `#!py|yaml` thing where Python generates YAML (because Jinja  in `salt-cloud`doesn't have access to many things, otherwise I would have used Jinja), so I'm not quite sure what to do with that. Switch to `#!py|jinja|yaml`? Looks scary.

And this is the first occurence. Python code looks like this:
```
    # Overrides for a single server (start with indent 2 here):
    srvconf['02'] = '''
  num_cpus: 4
  memory: 24GB
'''

... snip more Python code ...

    override = srvconf.get(vmset, "")
    ignore = "_ignore: &dry" if override else ""
    expand = "<<: *dry" if override else ""

    # This is the main profile YAML.
    obj = '''
{ignore}
{override}

{profile}:
  - infra-dbsnap-{vmset}.xdc:

      {expand}

      devices:
        network:
          Network adapter 1:
            ip: {dbsnap_ip}
...
'''.format(profile=profile, vmset=vmset,
           ignore=ignore, expand=expand, override=override, **ips)
```

Which is admittedly terribly ugly. The whole point of anchor is to be able to start with the known small amount of whitespace when specifying  `srvconf['02']`. Python could interpolate that string without anchors, but then a single typo could cause hours of debugging.
I hate all this and I'd really like to use Jinja instead of Python, but Jinja in salt-cloud's map files cannot use Salt's execution modules, so I don't see what else I could do.
Aside from the monstrosity above, Python code looks up environment variables (because I didn't manage to find any other mechanism to pass variables) and calls `socket.gethostbyname` to get IP addresses which then get interpolated.

**Comment by dkacar-oradian** [comment]
I just wrote all that crap yesterday to see if I can get it working. Looking at it now I have a few ideas to make it nicer and maybe even manageable. But the Python code would have to remove parts of the template which won't be top-level.

However, I think that anything which can save you from debugging whitespace problems is a worthy addition. So it would be nice if Salt's YAML processor could remove things starting with `_ignore` (or something configurable). Abusing that feature could lead to even more mess, of course.

**Comment by frogunder** [comment]
@saltstack/team-core any inputs / thoughts on this?

**Comment by dkacar-oradian** [comment]
I can just add that I rewrote my crap and now it has more Python code , looks a lot nicer and doesn't need this feature in Salt.
But I still think it might be a useful addition.
